File: graphs/max-area-of-island/sol.py
from typing import List


class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ans = 0
        visited = set()
        m, n = len(grid), len(grid[0])

        def dfs(i, j):
            if (
                i not in range(m)
                or j not in range(n)
                or grid[i][j] == 0
                or (i, j) in visited
            ):
                return 0
            else:
                visited.add((i, j))
                return 1 + dfs(i, j + 1) + dfs(i + 1, j) + dfs(i, j - 1) + dfs(i - 1, j)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1 and (i, j) not in visited:
                    ans = max(ans, dfs(i, j))

        return ans

File: graphs/max-area-of-island/sol.cpp
// https://leetcode.com/problems/max-area-of-island/description/

#include <vector>

using std::max;
using std::vector;

class Solution
{
public:
    int maxAreaOfIsland(vector<vector<int>> &grid)
    {
        int ans = 0;
        int m = grid.size();
        int n = grid.at(0).size();

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (grid.at(i).at(j) == 1)
                {
                    ans = max(ans, dfs(grid, i, j, m, n));
                }
            }
        }

        return ans;
    }

private:
    int dfs(vector<vector<int>> &grid, int i, int j, int m, int n)
    {
        if (i < 0 || i >= m || j < 0 || j >= n || grid.at(i).at(j) == 0)
        {
            return 0;
        }
        grid.at(i).at(j) = 0; // mark as visited
        return (1 + dfs(grid, i, j + 1, m, n) +
                dfs(grid, i + 1, j, m, n) +
                dfs(grid, i, j - 1, m, n) +
                dfs(grid, i - 1, j, m, n));
    }
};